src/std/io/bio/delimited.ss

```diff
@@ -55,6 +55,21 @@
           (bio-delimited-peek-u8 in)))
       '#!eof)))
 
+(def (bio-delimited-put-back delim previous-input)
+  (let* ((limit (&delimited-input-buffer-limit delim))
+         (remaining (&delimited-input-buffer-remaining delim))
+         (new-remaining
+          (fx+ (if (pair? previous-input) (length previous-input) 1)
+               remaining)))
+    (if (fx< new-remaining limit)
+      (let (in (&delimited-input-buffer-in delim))
+        (if (input-buffer? in)
+          (bio-put-back in previous-input)
+          (bio-delimited-put-back in previous-input))
+        (set! (&delimited-input-buffer-remaining delim)
+          new-remaining))
+      (raise-io-error 'BufferedReader-put-back "too many bytes returned" previous-input))))
```

Comment:
Does that mean that if I try to put back a character just at the wrong moment, I won't be able to? That sucks.

Reply:
IIRC, in CL, streams have a special field of a single optional unread character, that is checked on every read, to guarantee you can always unread one (and only one) character.

Reply:
Of course, with character vs bytes vs utf8 and giggles, you might have to support more than one character, and instead a small vector of bytes, or a deque, etc.

Reply:
Does that mean the delim variant should remember the start as well as the end of the limited range?

Reply:
it remembers the original limit, and it stops you from making remaing exceed the original limit.
So here it won't let you put back more than you read, pethaps we should relax this restriction.

Reply:
I removed the restriction so that it is consistent with the underlying buffer behavior and the interface specification.